liquibase/liquibase
# Fixed postgresql addPrimaryKey cluster handling

Fixes #1779 
[LB-1398]

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: [https://github.com/liquibase/liquibase/issues/1779](https://github.com/liquibase/liquibase/issues/1779)
* Local Branch: [https://github.com/liquibase/liquibase/tree/LB-1398](https://github.com/liquibase/liquibase/tree/LB-1398)
* Unit Tests: [https://github.com/liquibase/liquibase/pull/1880/checks](https://github.com/liquibase/liquibase/pull/1880/checks)
* Integration Tests: (shown in Unit Tests link above)
* Functional Tests: [https://jenkins.datical.net/job/liquibase-pro/job/LB-1398/](https://jenkins.datical.net/job/liquibase-pro/job/LB-1398/)

#### Testing
* Steps to Reproduce: [https://github.com/liquibase/liquibase/issues/1779](https://github.com/liquibase/liquibase/issues/1779)
* Guidance: 
  * Just impacts addPrimaryKey
  * Moved code for mssql logic, but should remain unchanged

#### Dev Verification
(Did not save output at the time)

## Test Requirements (Internal Liquibase QA)
**Manual Tests - Postgres**

Use the changelog attached to this ticket to run the tests on a Postgres database. 

_Verify update-sql does not throw a validation error for addPrimaryKey changes with_ `clustered=false`.

* `liquibase --output-file out.txt update-sql --changelog-file lb1779-changelog.yml`
* The output should **not** contain: clustered is not allowed on postgresql, lb1779-changelog.yml::1-nonClusteredPK::lb-1779

_Verify update-sql produces the correct statements to cluster primary keys._

|**CHANGESET ID**|**GENERATED SQL**|

|1-nonClusteredPK|ALTER TABLE public.tbl_noncluster ADD CONSTRAINT pk_nonclustered PRIMARY KEY (id);
There is no “CLUSTER” statement for this changeset.|
|2-clusteredPK|ALTER TABLE public.tbl_clustered ADD CONSTRAINT pk_clustered PRIMARY KEY (id);
CLUSTER public.tbl_clustered USING pk_clustered;|


|----|----|
|3-unnamedClusteredPK|Postgresql requires constraintName on addPrimaryKey when clustered=true|

|4-unnamedNonClusteredPK|ALTER TABLE public."tbl_namelessNonClusterPK" ADD PRIMARY KEY (id);
There is no ‘CLUSTER’ statement for this changeset.|

_Verify update is successful for a non-clustered primary key._

* `liquibase update --changelog-file lb1779-changelog.yml --labels setup,noncluster`

_Verify update is successful for a clustered primary key._

* `liquibase update --changelog-file lb1779-changelog.yml --labels cluster`

_Verify update returns an error when clustered is specified for an un-named primary key._

* `liquibase update --changelog-file lb1779-changelog.yml --labels missingNameValidation`
* Error message is: Postgresql requires constraintName on addPrimaryKey when clustered=true

**Manual Tests - SQL Server**

Use the attached changelog to run these tests. SQL Server code was moved by this change and we’re validating the clustered/nonclustered primary keys are correctly generated.

_Verify update-sql produces the correct statements to cluster primary key on SQL Server._

|**CHANGESET ID**|**GENERATED SQL**|
|----|----|
|1-nonClusteredPK|ALTER TABLE tbl_noncluster ADD CONSTRAINT pk_nonclustered PRIMARY KEY NONCLUSTERED (id)|
|2-clusteredPK|ALTER TABLE tbl_clustered ADD CONSTRAINT pk_clustered PRIMARY KEY CLUSTERED (id)|

* Note! SQL Server throws an error creating 3-unnamedClusteredPK. The error is: Cannot define PRIMARY KEY constraint on nullable column in table 'tbl_namelesspk'. 

_Verify update is successful for valid MSSQL table and primary key definitions._

* `liquibase update --changelog-file lb1779-changelog.yml --labels setup,noncluster,cluster`

**Automated Test Requirements**

* No new functional automated tests required.
* Two liquibase/liquibase-test-harness tests are required.
* Tests should run on Postgres.
  * Original requirement included MSSQL tests but that requirement is redacted based on code review and manual testing.
* Test Case 1
  * Add a changelog with changesets `id: 1-createTableForNonClusteredPK` and `1-nonClusteredPK`
    * liquibase-test-harness/src/main/resources/liquibase/harness/change/changelogs/
    * Ask a member of the test-harness team if they have a recommendation on naming and location of changelog.
  * Add the generated SQL for test validation:
    * src/main/resources/liquibase/harness/change/expectedSql/
  * Add the JSON snapshot for test validation:
    * src/main/resources/liquibase/harness/change/expectedSnapshot/
* Test Case 2
  * Add a changelog with changesets `id: 1-createTableForClusteredPK` and `2-clusteredPK`
    * liquibase-test-harness/src/main/resources/liquibase/harness/change/changelogs/
    * Ask a member of the test-harness team if they have a recommendation on naming and location of changelog.
  * Add the generated SQL for test validation:
    * src/main/resources/liquibase/harness/change/expectedSql/
  * Add the JSON snapshot for test validation:
    * src/main/resources/liquibase/harness/change/expectedSnapshot/



┆Issue is synchronized with this [Jira Story](https://datical.atlassian.net/browse/LB-1779) by [Unito](https://www.unito.io)


[LB-1398]: https://datical.atlassian.net/browse/LB-1398
